GTG/core/tasks2.py

```diff
@@ -0,0 +1,559 @@
+
+        if self.status is Status.ACTIVE:
+            self.status = Status.DONE
+            self.date_closed = Date.today()
+
+        else:
+            self.status = Status.ACTIVE
+            self.date_closed = Date.no_date()
+
+            if self.parent and self.parent.status is not Status.ACTIVE:
+                self.parent.toggle_status(propagate=False)
+
+        if propagate:
+            for child in self.children:
+                child.toggle_status()
+
+
+    def dismiss(self) -> None:
+        """Set this task to be dismissed."""
+
+        self.status = Status.DISMISSED
+
+        for child in self.children:
+            child.dismiss()
+
+    @property
+    def date_due(self) -> Date:
+        return self._date_due
+
+
+    @date_due.setter
+    def date_due(self, value: Date) -> None:
+        self._date_due = value
+
+        if not value or value.is_fuzzy():
+            return
+
+        for child in self.children:
+            if (child.date_due
+               and not child.date_due.is_fuzzy()
+               and child.date_due > value):
+
+                child.date_due = value
+
+        if (self.parent
+           and self.parent.date_due
+           and self.parent.date_due.is_fuzzy()
+           and self.parent.date_due < value):
+            self.parent.date_due = value
+
+
+    @property
+    def date_added(self) -> Date:
+        return self._date_added
+
+
+    @date_added.setter
+    def date_added(self, value: Any) -> None:
+        self._date_added = Date(value)
+
+
+    @property
+    def date_start(self) -> Date:
+        return self._date_start
+
+
+    @date_start.setter
+    def date_start(self, value: Any) -> None:
+        self._date_start = Date(value)
+
+
+    @property
+    def date_closed(self) -> Date:
+        return self._date_closed
+
+
+    @date_closed.setter
+    def date_closed(self, value: Any) -> None:
+        self._date_closed = Date(value)
+
+
+    @property
+    def date_modified(self) -> Date:
+        return self._date_modified
+
+
+    @date_modified.setter
+    def date_modified(self, value: Any) -> None:
+        self._date_modified = Date(value)
+
+
+    @property
+    def title(self) -> str:
+        return self.raw_title
+
+
+    @title.setter
+    def title(self, value) -> None:
+        self.raw_title = value.strip('\t\n') or _('(no title)')
+
+
+    @property
+    def excerpt(self) -> str:
+        if not self.content:
+            return ''
+
+        # Strip tags
+        txt = TAG_REGEX.sub('', self.content)
+
+        # Strip subtasks
+        txt = SUB_REGEX.sub('', txt)
+
+        # Strip blank lines and set within char limit
+        return f'{txt.strip()[:80]}…'
+
+
+    def add_tag(self, tag: Tag2) -> None:
+        """Add a tag to this task."""
+
+        if isinstance(tag, Tag2):
+            if tag not in self.tags:
+                self.tags.append(tag)
+
+                for child in self.children:
+                    child.add_tag(tag)
```

Comment:
Should we recursively add? For example, lets have a subtask about phone and a subsubtask like "prepare work". Now you add phone to the subtask and suddenly prepare work has also phone while you don't need it there.

Reply:
hmm that's a good question, I think it's mostly to keep consistency with the UI since tasks show up as inheriting tags from their parents in the list (same color, icon)

Reply:
Been thinking about this, and I think you are right. We can inherit tags when creating subtasks (as we do currently), and not recursively add/remove